Add option to seed LIGHT particles at cell-centers with gaussian noise added

bradyrx opened the pull request:
This PR adds a flag to the LIGHT particle initialization script to seed 3 particles at the cell-center with a Gaussian noise perturbation. It also allows a user to seed both the cell-center and vertices simultaneously.

bradyrx commented:
cc @pwolfram @kellykochanski. Per our discussions on our recent conference call.

This adds three flags to `make_particle_file`:
* `-c`/`--center`: Seed one particle at the cell-center. Defaults to true if no other flags are on.
* `-v`/`--vertex`: Seed three particles at the cell vertices by a fixed perturbation of CFLmin.
* `-n`/`--add_noise`: Adds Gaussian noise to the cell-center seeding to generate three particles seeded at each cell center.

Note that all three can be called now, so one can have off-vertex and cell-center seeding with Gaussian noise (for our scaling test use case).

Example usage:
`./make_particle_file.py -i mesh.nc -g graph.info.part.6 -p 6 -t surface -c -n -v -o gaussian_center_and_vertex.nc`

bradyrx commented:
An example particle initialization with @pwolfram's small mesh. This has the Gaussian noise turned on + off-vertex seeding. Easiest to view by opening this in a new window. You'll see that the off-vertex particles are a fixed epsilon from the vertices. For cell-center seeding, each cell has 3 unique directions they are perturbed in (of the possible 6) and a fixed Gaussian perturbation ranging from [CFLMin, 2*CFLMin].

Last note -- this script is in major need of a python3 makeover + refactoring. I'd be happy to take that on eventually, @pwolfram.

<img width="1877" alt="Screen Shot 2019-06-06 at 10 47 29 AM" src="https://user-images.githubusercontent.com/8881170/59051392-038c6200-884a-11e9-8d76-f94985f7f8ae.png">

pwolfram commented on testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py:
```diff
@@ -348,57 +347,105 @@ def write(self, f_name, f_decomp):  #{{{
+    
+    # assumes particle moves more than 1/100 of a cell scale per time step
+    # 10000 m cell, 0.1m/s flow, 2 hr time step is CFL=0.07
+    CFLmin = 0.005
```
Move as default argument?

pwolfram commented on testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py:
```diff
@@ -579,11 +630,18 @@ def build_particle_file(f_init, f_name, f_decomp, types, spatialfilter, buoySurf
+        args.seed_center = True
+
+    if args.add_noise and not args.seed_center:
+        raise ValueError('Gaussian noise requested but center-seeding not requested. Please resubmit this function with `--center` and `--add_noise`.')
```
Please split into multiple lines.

bradyrx replied:
Done. Will convert this to python3 down the line and will run `black` on this for syntax.

pwolfram commented on testing_and_setup/compass/utility_scripts/LIGHTparticles/make_particle_file.py:
```diff
@@ -484,9 +531,9 @@ def build_surface_floats(cpts, xCell, yCell, zCell, afilter): #{{{
 
 
-def build_particle_file(f_init, f_name, f_decomp, types, spatialfilter, buoySurf, nVertLevels, downsample, vertseedtype, loc): #{{{
+def build_particle_file(f_init, f_name, f_decomp, types, spatialfilter, buoySurf, nVertLevels, downsample, vertseedtype, seed_center, seed_vertex, add_noise): #{{{
```
Please split line.

bradyrx commented:
Okay @pwolfram, this is good to merge. I rebased and incorporated your edits. I also tested the various argument combinations and they work fine. 

Base argument example (automatically seeds cell-centers with one particle):
`./make_particle_file.py -i mesh.nc -g graph.info.part.6 -p 6 -o output.nc` 

Use just `-v` for off-vertex seeding. Use `-v -c` for off-vertex and center-seeding. Use `-c -n` for center-seeding with three particles separated via Gaussian noise. Our LIGHT methods paper will use `-c -n -v` to have off-vertex seeding plus center-seeding with Gaussian noise.

mark-petersen commented:
Looks like this is ready to merge, based on @pwolfram approval and comments above. @bradyrx Thanks for helping us out.

pwolfram commented:
Thanks for the merge @mark-petersen !
